feat: support collapse a single folder under cursor

linepi opened the pull request:
When I use `api.tree.expand_all` to expand a folder, sometimes I need to collapse a single folder rather than all folders(invoke by `api.tree.collapse_all`). This feature(`api.tree.collapse`) can be understood as the inverse operation of `api.tree.expand_all`.

linepi commented:
There might be some imperfections in the naming, documentation, and code details. What I am presenting here is just a rough idea. If this pull request idea is accepted, I will continue to work hard to make more reasonable, reliable, and complete revisions.

evertonse commented:
> When I expand a folder, sometimes I need to collapse a single folder rather than all folders. This can be understood as the inverse operation of expanding.

I think I migh not understand what you mean. 
Is this ``api.node.navigate.parent_close`` *not* what you want?

linepi commented:
@evertonse No, that is not the same. api.node.navigate.parent_close will close the parent folder of where cursor is. Collapse will close folder and all it subfolder recursively.

evertonse commented:
> @evertonse No, that is not the same. `api.node.navigate.parent_close` will close the parent folder of where cursor is. The new feature `api.tree.collapse` will close folder and all it subfolder recursively.

Ah I see now, then there should be ``api.tree.collapse``. Have at it

linepi commented:
I update my commit, including the command, doc and explanation. Please review.

alex-courtis commented on doc/nvim-tree-lua.txt:
```diff
@@ -1770,6 +1778,9 @@ tree.collapse_all({keep_buffers})          *nvim-tree-api.tree.collapse_all()*
     Parameters: ~
       • {keep_buffers} (boolean) do not collapse nodes with open buffers.
 
+tree.collapse()          *nvim-tree-api.tree.collapse()*
```
```suggestion
tree.collapse()                                *nvim-tree-api.tree.collapse()*
```
spacing

alex-courtis reviewed (changes requested):
Apologies for not seeing this one earlier...

This is great, we should have had this long ago. I'm thinking about a free default keybinding we can use.

We will do this this however:
* It's missing the `keep_buffers` parameter of `collapse_all`
* `collapse.lua` repeats `collapse-all.lua` taking a node instead of using `explorer`

Please an optional node parameter to `collapse-all.fn` so that it may perform both and keep buffers.

alex-courtis commented:
This is great, many thanks for pushing this one.

linepi commented:
> Please an optional node parameter to `collapse-all.fn` so that it may perform both and keep buffers.

All right, I've added and option to `collapse_all.fn` and removed `collapse`. Now it's only one function. Please review.

alex-courtis commented:
On `:NvimTreeCollapseKeepBuffers`

```
Error executing Lua callback: ...pi/lua/nvim-tree/actions/tree/modifiers/collapse-all.lua:29: attempt to index local 'opts' (a boolean value)
stack traceback:
        ...pi/lua/nvim-tree/actions/tree/modifiers/collapse-all.lua:29: in function 'collapse_all'
        ...site/pack/packer/start/linepi/lua/nvim-tree/commands.lua:145: in function <...site/pack/packer/start/linepi/lua/nvim-tree/commands.lua:144>
```

alex-courtis commented:
```help
tree.collapse_all({keep_buffers})          *nvim-tree-api.tree.collapse_all()*
    Collapse the tree.

    Parameters: ~
      • {keep_buffers} (boolean) do not collapse nodes with open buffers.

tree.collapse({opts})                          *nvim-tree-api.tree.collapse()*
    Collapse the tree at the node under the cursor.

    Parameters: ~
      • {opts} (table) optional parameters

    Options: ~
      • {keep_buffers} (boolean) do not collapse nodes with open buffers,
                                 default false
```

alex-courtis reviewed (changes requested):
Looking good, we are close...

alex-courtis commented:
Closing as this PR is stale. Please reopen if you wish to continue this work.
